File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.js
const readline = require("readline");

let n;
let numList;
let operList;
let max = -Number.MAX_VALUE;
let min = Number.MAX_VALUE;

const selectOper = (result, targetIndex) => {
    if (targetIndex === n) {
        max = Math.max(max, result);
        min = Math.min(min, result);
        return;
    }

    if (operList[0] > 0) {
        operList[0]--;
        selectOper(result + numList[targetIndex], targetIndex + 1);
        operList[0]++;
    }
    if (operList[1] > 0) {
        operList[1]--;
        selectOper(result - numList[targetIndex], targetIndex + 1);
        operList[1]++;
    }
    if (operList[2] > 0) {
        operList[2]--;
        selectOper(result * numList[targetIndex], targetIndex + 1);
        operList[2]++;
    }
    if (operList[3] > 0) {
        operList[3]--;
        const temp = parseInt(result / numList[targetIndex]);
        selectOper(temp ? temp : 0, targetIndex + 1);
        operList[3]++;
    }
};

const inputList = [];
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
});
rl.on("line", (line) => {
    inputList.push(line);

    if (inputList.length === 3) {
        rl.close();
    }
}).on("close", () => {
    n = Number(inputList[0]);
    numList = inputList[1].split(" ").map(Number);
    operList = inputList[2].split(" ").map(Number);

    selectOper(numList[0], 1);

    console.log(max);
    console.log(min);

    process.exit();
});

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
n = int(input())
num_list = list(map(int, input().split()))
oper_list = list(map(int, input().split()))

max_result = -float("inf")
min_result = float("inf")


def select_operator(result, index):
    global max_result, min_result

    if index == n:
        max_result = max(max_result, result)
        min_result = min(min_result, result)
        return

    if oper_list[0] > 0:
        oper_list[0] -= 1
        select_operator(result + num_list[index], index + 1)
        oper_list[0] += 1
    if oper_list[1] > 0:
        oper_list[1] -= 1
        select_operator(result - num_list[index], index + 1)
        oper_list[1] += 1
    if oper_list[2] > 0:
        oper_list[2] -= 1
        select_operator(result * num_list[index], index + 1)
        oper_list[2] += 1
    if oper_list[3] > 0:
        oper_list[3] -= 1
        select_operator(int(result / num_list[index]), index + 1)
        oper_list[3] += 1


select_operator(num_list[0], 1)
print(max_result)
print(min_result)